### ISSUE TITLE
Icon files: ICO, EXE

### ISSUE THREAD

**Issue by GerbilSoft** [GerbilSoft/rom-properties#170]
Add support for icon files:
* ~~ICNS: Mac OS~~
* ICO: Windows
* EXE: Primary ICO format icon for executables

NOTE: EXE parsing probably has to be disabled on Windows. Note that newer versions of Windows don't support extracting icons from 16-bit executables, so it might be worth seeing if that can somehow be done anyway.

Also, maybe add icon support for Mac application directories and/or DMGs. DMG support will require file system parsing and decompression.

This enhancement depends on #67: Add support for multiple image sizes. (internal images)

**Comment by baka0815** [comment]
Are you sure you want to implement that? Doesn't every file browser (dolphin, thunar nautilus,...) display those already?

**Comment by GerbilSoft** [comment]
I did check Dolphin, and it does seem to have thumbnailing for these formats. Recent Dolphin updates improved Windows icon thumbnailing (previously it was using small icons), but ICNS is still showing up small.

ICNS is definitely not supported on Windows though, so it'd benefit from that.

Also, Dolphin doesn't support thumbnailing DMGs. Granted, DMG thumbnailing is a lot more difficult than just ICNS thumbnailing...

**Comment by GerbilSoft** [comment]
ICO/EXE is implemented, at least for the "best" size.

Going to split ICNS into a separate issue.

**Comment by kristibektashi** [comment]
@GerbilSoft about the 16-bit executable icons, yes it can be done, and in fact it has already been done: https://github.com/otya128/Icon16bitFix

**Comment by GerbilSoft** [comment]
Looks like that works by detouring PrivateExtractIconsW() and extracting 16-bit icons manually if the regular PrivateExtractIconsW() function doesn't work. I'll look into doing something like that.

If I can implement it, it will be made configurable in rp-config as well.

**Comment by GerbilSoft** [comment]
Added in a new branch: https://github.com/GerbilSoft/rom-properties/tree/feature/win16-icons-on-win64

Only minor issue is that it won't thumbnail Win16 executables if the rom-properties shell extension hasn't been loaded yet. (I'm not setting rom-properties as the .exe icon handler because that could cause problems.)

I'm using the same Detours method as Icon16bitFix, but without the resident program hooking processes. Presumably that's what caused false-positive virus warnings with the original.

**Comment by kristibektashi** [comment]
Btw I’m pretty sure that even recent 32-bit versions of Windows 10 don’t show 16-bit icons nowadays so the name of the branch is inaccurate. But I might be wrong about this, this needs more testing on 32-bit Windows
